src/extension.ts

```diff
@@ -20,7 +31,123 @@ export async function activate(context: vscode.ExtensionContext) {
+			}
+			const activeTextEditor = vscode.window.activeTextEditor;
+			if (activeTextEditor !== undefined) {
+				pendingFooJsonDecoration = setTimeout(() => updateFooJsonDecorations(activeTextEditor), 1000);
+			}
+		}
+	}, null, context.subscriptions));
+
+	// decorate the active editor now
+	const activeTextEditor = vscode.window.activeTextEditor;
+	if (activeTextEditor !== undefined) {
+		updateFooJsonDecorations(activeTextEditor);
+	}
+
+	// decorate when then cursor moves
+	context.subscriptions.push(new EditorEventHandler());
 }
 
 // This method is called when your extension is deactivated
 export function deactivate() { }
+
+const argDecoration = vscode.window.createTextEditorDecorationType({
+	light: {
+		color: '#ff6f00'
+	},
+	dark: {
+		color: '#fff9c4'
+	}
+});
+const selectDecoration = vscode.window.createTextEditorDecorationType({
+	light: {
+		color: '#6a1b9a'
+	},
+	dark: {
+		color: '#ce93d8'
+	}
+});
+const pluralDecoration = vscode.window.createTextEditorDecorationType({
+	light: {
+		color: '#0277bd'
+	},
+	dark: {
+		color: '#b3e5fc'
+	}
+});
+
+const selectRegex = /^(\w+\s*,\s*select\s*,(?:\s*\w+\{.*\})*)$/g;
+const pluralRegex = /^(\w+\s*,\s*plural\s*,(?:\s*\w+\{.*\})*)$/g;
+function updateFooJsonDecorations(editor: vscode.TextEditor) {
+	if (!editor || !path.basename(editor.document.fileName).endsWith('.arb')) {
+		return;
+	}
+	let colorMap = new Map<string, vscode.Range[]>();
+	visit(editor.document.getText(), {
+		onLiteralValue: (value: string, offset: number) => {
+			let openBrackets: number[] = [];
+			for (let index = 0; index < value.length; index++) {
+				const char = value.charAt(index);
+				if (char === '{') {
+					openBrackets.push(index);
+				} else if (char === '}') {
+					let start = openBrackets.pop() ?? 0;
+					let end = index;
+					const part = value.substring(start + 1, end);
+					const level = openBrackets.length;
+					const decoration = parse(part, level);
+					if (decoration !== undefined) {
+						const rangeStart = offset + start + 1;
+						const rangeEnd = offset + end + 2;
+						colorMap.set(decoration, [...(colorMap.get(decoration) ?? []), new vscode.Range(editor.document.positionAt(rangeStart), editor.document.positionAt(rangeEnd))]);
+					}
+				}
+			}
+		}
+	});
+	editor.setDecorations(pluralDecoration, colorMap.get('plural') ?? []);
+	editor.setDecorations(selectDecoration, colorMap.get('select') ?? []);
+	editor.setDecorations(argDecoration, colorMap.get('arg') ?? []);
+}
+
+function parse(value: string, level: number): string | undefined {
+	if (selectRegex.exec(value) !== null) {
+		return 'select';
+	} else if (pluralRegex.exec(value) !== null) {
+		return 'plural';
+	} else {
+		return 'arg';
+	}
+}
+
+class EditorEventHandler {
+	private _disposable: vscode.Disposable;
+
+	constructor() {
+		let subscriptions: vscode.Disposable[] = [];
+		vscode.window.onDidChangeTextEditorSelection((e: vscode.TextEditorSelectionChangeEvent) => {
```

Comment:
Why does this need to be run on cursor moves? (I presume `onDidChangeTextDocument` is taking care of actual document changes)

Reply:
It doesn't... :)